<a href="https://colab.research.google.com/github/YitbarekWedajo/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
### YOUR WORK HERE
import numpy as np
import pandas as pd
column_header = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style',
                 'drive-wheels', 'engine-location', 'wheel-base', ' length', 'width', 'height', 'curb-weight', 
                 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', ' compression-ratio', 
                 'horsepower', 'peak-rpm', 'city-mpg', ' highway-mpg', 'price']

In [2]:
Automobile=pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=column_header, na_values="?")
Automobile.head()

symboling  normalized-losses         make  ... city-mpg  highway-mpg    price
0          3                NaN  alfa-romero  ...       21           27  13495.0
1          3                NaN  alfa-romero  ...       21           27  16500.0
2          1                NaN  alfa-romero  ...       19           26  16500.0
3          2              164.0         audi  ...       24           30  13950.0
4          2              164.0         audi  ...       18           22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [3]:
### YOUR WORK HERE
#generate sample using sample()
sample = Automobile.sample(10, random_state=30)
sample.head(10)
sample_df=sample[' highway-mpg']

In [4]:
#calculate the mean of highway-mpg for the entire dataset 
Automobile[' highway-mpg'].mean()

30.75121951219512

In [5]:
#calculate the mean of highway-mpg for the sample dataset
sample[' highway-mpg'].mean()

32.8

In [6]:
#from calculated mean for the sample and the entire dataset, it is clear the two values are different. 

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [7]:
### YOUR WORK HERE

### 1) Null Hypothesis

The hypothesis that we might "nullify"

$H_0: \overline{x} == \mu$

My sample mean is equal to the population.
The cars sold have similar fuel efficency (35 miles per gallon)



2) Alternative Hypothesis
$H_a: \overline{x} \neq \mu$
The cars have differnt fuel efficency 

### 3) Confidence Level: 

Confidence Level: 95%

In [8]:
from scipy import stats

In [9]:
stats.ttest_1samp(sample_df, 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

4) Conclusion:

Based on a t-statistic of -0.95 and a p-value of 0.36 (reject/**fail to reject**) the null hypothesis that The cars sold have similar fuel efficency (35 miles per gallon), and suggest the alternative, cars sold have different fuel efficency.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

In [10]:
#with sample size 100
sample2=Automobile.sample(100, random_state=30)
sample2_df=sample2[' highway-mpg']

In [11]:
stats.ttest_1samp(sample2_df, 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

Based on a t-statistic of -6.51 and a p-value of ~0 (**reject**/fail to reject) the null hypothesis that The cars sold have similar fuel efficency (35 miles per gallon), and suggest the alternative, cars sold have different fuel efficency. The more the sample size is, the more likely close to true value.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [12]:
### YOUR WORK HERE
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column_headers, na_values="?")

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [13]:
### YOUR WORK HERE
#perform Categorical Encoding encoding 
df = df.replace({'y': 1, 'n': 0})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [14]:
# Compare republicans to democrats

rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

In [15]:
#calculate the mean for physician-fee-freeze freeze for both datasets
print(rep['physician-fee-freeze'].mean())
dem['physician-fee-freeze'].mean()

0.9878787878787879


0.05405405405405406

1) Null Hypothesis:

Democrats support this bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$[link text](https://)

2) Alternative Hypothesis:

Democrats support this bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 99%

In [16]:
#pass in the datasets of interest to calculate p values. NaN values has to be dropped as it could affect data size
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

4. Due to a t-statistic of 49.36 and a p-value of ~0 we **reject**/fail to reject the null hypothesis that democrats and repulicans support the physician-fee-freeze bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties. more Repblicans (98.78%) supported this bill as oppsed to only 50% of Democrats.

In [17]:
print(dem['budget'].mean())
rep['budget'].mean()

0.8884615384615384


0.13414634146341464

1) Null Hypothesis:

Democrats support budget bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis:

Democrats support budget bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 99%

In [18]:
stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

4. Due to a t-statistic of 49.36 and a p-value of ~0 we **reject**/fail to reject the null hypothesis that democrats and repulicans support the budget bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties. More Democrats (88.8 %%) supported this bill as oppsed to only 13.4 of Republicans.

In [23]:
#Find an issue where the difference between republicans and democrats has p > 0.1 
#(Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)
print(dem['immigration'].mean())
rep['immigration'].mean()

0.4714828897338403


0.5575757575757576

1) Null Hypothesis:

Democrats support immigration bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis:

Democrats support budget immigration bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 90%

In [24]:
stats.ttest_ind(dem['immigration'], rep['immigration'], nan_policy='omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

Due to a t-statistic of -1.73 and a p-value of 0.083 we reject/**fail to reject** the null hypothesis that democrats and repulicans support the immigration bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.In fact the two parties support this bill with no significance difference( Republicans 55.55% and Democrats 47.14%)

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE